File: core/src/ca/kanoa/zombieworld/events/EventManager.java
package ca.kanoa.zombieworld.events;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.TreeSet;

public class EventManager {

    private HashMap<Class<? extends Event>, TreeSet<EventHandlerContainer>> handlers;
    private EventHandlerComparator comparator;

    public EventManager() {
        handlers = new HashMap<Class<? extends Event>, TreeSet<EventHandlerContainer>>();
        comparator = new EventHandlerComparator();
    }

    public void register(EventListener handler) throws EventRegistrationExeception {
        for (Method method : handler.getClass().getMethods()) {
            if (method.getAnnotation(EventHandler.class) != null) {
                // check for errors in event handler method
                if (method.getParameterTypes().length != 1) {
                    throw new EventRegistrationExeception(handler, "invalid number of parameters, should be 1");
                } else if (!Event.class.isAssignableFrom(method.getParameterTypes()[0])) {
                    throw new EventRegistrationExeception(handler, "handler parameter is not of the event type");
                } else {
                    registerEvent((Class<? extends Event>) method.getParameterTypes()[0], method, handler);
                }
            }
        }
    }

    private void registerEvent(Class<? extends Event> event, Method method, EventListener listener) {
        if (!handlers.containsKey(event)) {
            handlers.put(event, new TreeSet<EventHandlerContainer>(comparator));
        }
        handlers.get(event).add(new EventHandlerContainer(method, listener));
    }

    public int dispatchEvent(Event event) {
        if (handlers.containsKey(event.getClass())) {
            int failures = 0;
            TreeSet<EventHandlerContainer> eventHandlers = handlers.get(event.getClass());
            for (EventHandlerContainer eventHandler : eventHandlers) {
                try {
                    eventHandler.getMethod().invoke(eventHandler.getListener(), event);
                } catch (Exception e) {
                    e.printStackTrace();
                    failures++;
                }
            }
            return failures;
        } else {
            return -1;
        }
    }

    protected static class EventHandlerContainer {
        private Method method;
        private EventListener listener;

        private EventHandlerContainer(Method method, EventListener listener) {
            this.method = method;
            this.listener = listener;
        }

        protected Method getMethod() {
            return method;
        }

        protected EventListener getListener() {
            return listener;
        }
    }

}

File: core/src/ca/kanoa/zombieworld/events/EventManager.kt
package ca.kanoa.zombieworld.events

import java.lang.reflect.Method
import java.util.*

class EventManager {

    private val handlers: HashMap<Class<out Event>, TreeSet<EventHandlerContainer>>
    private val comparator: EventHandlerComparator

    init {
        handlers = HashMap<Class<out Event>, TreeSet<EventHandlerContainer>>()
        comparator = EventHandlerComparator()
    }

    @Throws(EventRegistrationExeception::class)
    fun register(handler: EventListener) {
        for (method in handler.javaClass.methods) {
            if (method.getAnnotation(EventHandler::class.java) != null) {
                // check for errors in event handler method
                if (method.parameterTypes.size != 1) {
                    throw EventRegistrationExeception(handler, "invalid number of parameters, should be 1")
                } else if (!Event::class.java.isAssignableFrom(method.parameterTypes[0])) {
                    throw EventRegistrationExeception(handler, "handler parameter is not of the event type")
                } else {
                    registerEvent(method.parameterTypes[0] as Class<out Event>, method, handler)
                }
            }
        }
    }

    private fun registerEvent(event: Class<out Event>, method: Method, listener: EventListener) {
        if (!handlers.containsKey(event)) {
            handlers.put(event, TreeSet(comparator))
        }
        handlers[event]!!.add(EventHandlerContainer(method, listener))
    }

    fun dispatchEvent(event: Event): Int {
        if (handlers.containsKey(event.javaClass)) {
            var failures = 0
            val eventHandlers: Set<EventManager.EventHandlerContainer> = handlers[event.javaClass] ?: Collections.emptySet()
            for (eventHandler in eventHandlers) {
                try {
                    eventHandler.method.invoke(eventHandler.listener, event)
                } catch (e: Exception) {
                    e.printStackTrace()
                    failures++
                }

            }
            return failures
        } else {
            return -1
        }
    }

    class EventHandlerContainer constructor(val method: Method, val listener: EventListener)

}
